Made changes needed to include a chicane benchmark.

cemitch99 opened the pull request:
1) Added a new "dipedge" element to correcty include dipole
   edge focusing

2) Corrected treatment of momenta in elements (affects only the dipole case)

3) Added a new input file "input_chicane.in" for compression of
   a beam in the Berlin-Zeuthen benchmark chicane.

cemitch99 committed:
Made changes needed to include a chicane benchmark.

1) Added a new "dipedge" element to correcty include dipole
   edge focusing

2) Corrected treatment of momenta in elements (affects only the dipole case)

3) Added a new input file "input_chicane.in" for compression of
   a beam in the Berlin-Zeuthen benchmark chicane.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+            // access AoS data such as positions and cpu/id
+            amrex::ParticleReal const x = p.pos(0);
+            amrex::ParticleReal const y = p.pos(1);
+            amrex::ParticleReal const t = p.pos(2);
```
Unused variable:
```suggestion
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+            // apply edge focusing
+            px = px + R21*x;
+            py = py + R43*y;
+
```
```suggestion
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+                amrex::ParticleReal & AMREX_RESTRICT px,
+                amrex::ParticleReal & AMREX_RESTRICT py,
+                amrex::ParticleReal & AMREX_RESTRICT pt,
+                RefPart const refpart) const {
```
To silence the unused variable warning without changing the interface
```suggestion
                [[maybe_unused]] RefPart const refpart) const {
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         * @param refpart reference particle
```
```suggestion
         * @param refpart reference particle (unused)
```

ax3l replied:
This is curious: why don't we need the reference particle for this implementation when we do need it for the simple sbend?

cemitch99 replied:
This model of the edge focusing depends only on the rotation angle of the magnet pole face and the magnet radius of curvature; it does not depend on the reference energy.  However, other more sophisticated models of the pole face rotation and the dipole fringe field could require the reference values.  That is why I left this as an input argument.

ax3l commented on src/particles/elements/Drift.H:
```diff
@@ -53,17 +53,27 @@ namespace impactx
+            // ptout = pt;
+
+            // assign updated momenta
+            px = pxout;
+            py = pyout;
+            pt = ptout;
```
I cannot see where we change the values for `px` or `pxout` in this function (or `y` or `t`), is this really needed for the drift?

cemitch99 replied:
No, this is not needed for the drift.  I made this change so that the logic of all of the elements is the same.  If someone uses an existing element as the template for a new one, then this could avoid confusion.

ax3l replied:
I see. We have to check if the compiler identifies and removes those extra copies.

Otherwise, we might want to remove the unneeded ones and/or create a custom element that we want to use as template?

ax3l replied:
Discussed today:
- current structure is nice because if we comment the code for momentum in between, we keep the identity map (for playing/debugging/new elements)
- the compiler recognized the overwritten initial assignment and discards it in any optimization mode >= `-O1`: https://godbolt.org/z/1Ecnr53oE

ax3l commented on src/particles/elements/Sbend.H:
```diff
@@ -54,23 +54,38 @@ namespace impactx
             amrex::ParticleReal const t = p.pos(2);
 
+            // initialize output values of momenta
+            amrex::ParticleReal pxout = px;
+            amrex::ParticleReal pyout = py;
+            amrex::ParticleReal ptout = pt;
```
I think this is only needed for `px`, not for `py` and `pt` in the sbend.

cemitch99 replied:
Yes, that's true.  See the previous comment.

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+beam.distribution = waterbag
+beam.sigmaX = 2.2951017632e-5
+beam.sigmaY = 1.3084093142e-5
+#beam.sigmaT = 5.5555553e-8
```
Maybe just leftover:
```suggestion
```

cemitch99 replied:
I tried two sets of parameters for the longitudinal phase space.  Let me verify which is best, and I'll delete the second and push the change.

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+beam.sigmaT = 6.00800897618671048624845e-8
+beam.sigmaPx = 1.598353425e-6
+beam.sigmaPy = 2.803697378e-6
+#beam.sigmaPt = 2.000000000e-6
```
Maybe just leftover:
```suggestion
```

cemitch99 replied:
Same comment as above.

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+beam.sigmaPt = 2.1628993476497387730437553e-6
+beam.muxpx = 0.933345606203060
+beam.muypy = 0.933345606203060
+#beam.mutpt = 0.999999961419755
```
Maybe just leftover:
```suggestion
```

cemitch99 replied:
Same comment as above.

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+                   dipedge2 drift1 dipedge1 sbend1 drift3
+
+sbend1.type = sbend
+sbend1.ds = 0.50037       #projected length 0.5 m, angle 2.77 deg
```
a bit more common/easier to read:
```suggestion
sbend1.ds = 0.50037       # projected length 0.5 m, angle 2.77 deg
```

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+sbend1.rc = -10.35
+
+drift1.type = drift
+drift1.ds = 5.0058489435  #projected length 5 m
```
a bit more common/easier to read:
```suggestion
drift1.ds = 5.0058489435  # projected length 5 m
```

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+drift1.ds = 5.0058489435  #projected length 5 m
+
+sbend2.type = sbend
+sbend2.ds = 0.50037       #projected length 0.5 m, angle 2.77 deg
```
a bit more common/easier to read:
```suggestion
sbend2.ds = 0.50037       # projected length 0.5 m, angle 2.77 deg
```

ax3l commented on examples/input_chicane.in:
```diff
@@ -0,0 +1,64 @@
+drift3.type = drift
+drift3.ds = 2.0
+
+dipedge1.type = dipedge   #dipole edge focusing
```
a bit more common/easier to read:
```suggestion
dipedge1.type = dipedge   # dipole edge focusing
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+/* Copyright 2021 Chad Mitchell, Axel Huebl
```
```suggestion
/* Copyright 2021-2022 Chad Mitchell, Axel Huebl
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,88 @@
+    {
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** Edge focusing associated with bend entry or exit
```
Awesome! Don't forget to add in the user-facing docs please:
- https://impactx.readthedocs.io/en/latest/usage/parameters.html#lattice-elements
- aka: `docs/source/usage/parameters.rst`

cemitch99 replied:
Upcoming in next commit.

ax3l reviewed (commented):
Brilliant, thanks a lot!

I added a first review and looked what CI suggests in the failing tests. Please see inline :)

cemitch99 committed:
Apply suggestions from code review

Applying initial suggestions from Axel for #75 (excluding input comment lines).

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Added documentation of "dipedge" and set "input_chicane.in" values.

cemitch99 committed:
Merged Axel's committed suggestions.

Merge branch 'benchmark_chicane' of https://github.com/cemitch99/impactx into benchmark_chicane

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,86 @@
+                PType& AMREX_RESTRICT p,
+                amrex::ParticleReal & AMREX_RESTRICT px,
+                amrex::ParticleReal & AMREX_RESTRICT py,
+                amrex::ParticleReal & AMREX_RESTRICT pt,
```
```suggestion
                [[maybe_unused]] amrex::ParticleReal & AMREX_RESTRICT pt,
```

ax3l committed:
DipEdge: Unused parameter warning (pt)

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,86 @@
+         * @param p Particle AoS data for positions and cpu/id
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
```
```suggestion
         * @param pt particle momentum in t (unchanged)
```

ax3l committed:
Update pt comment: unchanged

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,86 @@
+            // access reference particle values if needed
+
+            // edge focusing matrix elements (zero gap)
+            amrex::ParticleReal R21 = tan(m_psi)/m_rc;
+            amrex::ParticleReal R43 = -tan(m_psi)/m_rc;
```
Tiny improvement since `tan` is relatively expensive and this makes sure we only compute it once:
```suggestion
            amrex::ParticleReal const R21 = tan(m_psi)/m_rc;
            amrex::ParticleReal R43 = -R21;
```

cemitch99 replied:
Great, except the updated line 64 should read R43 = -R21.

ax3l replied:
Great catch, updated now :)

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,86 @@
+            // first-order effect of nonzero gap
+            vf = (1.0_prt + pow(sin(m_psi),2))/(pow(cos(m_psi),3));
+            vf = vf * m_g * m_K2/(pow(m_rc,2));
+            R43 = R43 + vf;
```
We could write this as follows, if we want to. Same meaning, same cost:
```suggestion
            vf *= m_g * m_K2/(pow(m_rc,2));
            R43 += vf;
```

ax3l commented on src/particles/elements/DipEdge.H:
```diff
@@ -0,0 +1,86 @@
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** Edge focusing associated with bend entry or exit
+         *
```
Hi @cemitch99, does it make sense to add the reference of the model you used here?
```suggestion
         *
         * This model assumes a first-order effect of nonzero gap.
         * Here we use the linear fringe field map, given to first order in g/rc (gap / radius of curvature).
         *
         * References:
         *   K. L. Brown, SLAC Report No. 75 (1982).
         *   K. Hwang and S. Y. Lee, PRAB 18, 122401 (2015).
         *
```

cemitch99 replied:
Yes, I would add two references:

K. L. Brown, SLAC Report No. 75, 1982.
K. Hwang and S. Y. Lee, PRAB 18, 122401 (2015).

Here we use the linear fringe field map, given to first order in g/rc (gap / radius of curvature).

ax3l replied:
Perfect, added :)

cemitch99 committed:
Update src/particles/elements/DipEdge.H

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l committed:
Add Reference & Spare extra tan()

ax3l reviewed (approved):
Looks splendid, in it goes :)
